File: leetcode_3163/StringCompressionIII.java
package leetcode_3163;


class StringCompressionIII{
    public String compressedString(String word) {
        StringBuilder res = new StringBuilder();
        int count = 1;
        // we add the first character to the result
        char ch = word.charAt(0);

        // loop through the word and count the number of the same character
        // we start from the second character as we have already added the first character to the result
        // this is because we need to compare the current character with the previous one
        for(int i = 1; i < word.length(); i++){
            // if the character is the same as the previous one, increase the count
            if(word.charAt(i) == ch && count < 9)
            {
                count++;
            }
            // if the character is different from the previous one, append the count and the character to the result
            // then reset the count to 1
            // and update the character to the current character
            else
            {
                res.append(count).append(ch);
                count = 1;
                ch = word.charAt(i);
            }
        }
        // append the last count and character to the result
        // this is because the loop above will not add the last character
        res.append(count).append(ch);

        return res.toString();
    }
    
    public static void main(String[] args) {
        StringCompressionIII sc = new StringCompressionIII();
        System.out.println(sc.compressedString("abcde"));
        System.out.println(sc.compressedString("aaaaaaaaaaaaaabb"));
    }
}

File: leetcode_3163/StringCompressionIII.py
class StringCompressionIII:
    def compressedString(self, word: str) -> str:
        res = ""
        count = 1
        # we add the first character to the result
        ch = word[0]
        
        # we iterate through the word and count the number of times a character appears 
        for i in range(1, len(word)):
            # if the character is the same as the previous one and its count is less then 9
            # then we increment the count 
            if word[i] == ch and count < 9:
                count += 1
            # if the character is different from the previous one 
            # then we add its count and the previous character to the result
            # and reset the count to 1
            # and update the character to the current character
            else:
                res += str(count) + ch
                count = 1
                ch = word[i]
                
        # we add the last character and its count to the result
        # this is because the loop above will not add the last character
        res += str(count) + ch
        return res

def main():
    word1 = "abcde"
    word2 = "aaaaaaaaaaaaaabb"
    sc = StringCompressionIII()
    print(sc.compressedString(word1))
    print(sc.compressedString(word2))

if __name__ == "__main__":
    main()